### PR TITLE
Add support for Join/Leave methods to Endpoint

### PR DESCRIPTION
**Pull request by mrjana**
- Added support for Join/Leave methods to Endpoint.
- Removed sandbox key argument from CreateEndpoint.
- Refactored bridge driver code to remove sandbox key.
- Fixed bridge driver code for gaps in ipv6 behavior
  observed during docker integration.
- Updated test code, readme code, README.md according
  api change.
- Fixed some sandbox issues while testing docker ipv6
  integration.

Signed-off-by: Jana Radhakrishnan <mrjana@docker.com>

### CODE REVIEW

**Comment by squaremo** [comment]
Is the motivation here for endpoints to have a lifetime independent of being attached to a container, or something more innocuous?

**Comment by mrjana** [comment]
@squaremo The motivation here is actually the former. As mentioned earlier in other threads, the idea is to keep the endpoint lifetime independent of containers. Then endpoint actually provides the abstraction of a service which is advertised on the network and the container the actual provider of the service. This separation facilitates different containers attached to the same endpoint at different times providing the service. This can be completely transparent to the client. The end-to-end implementation (in terms of UI and such) are not available yet but we would like to settle down on the api pretty early.

File: sandbox/namespace_linux.go [hunk 1]
```diff
@@ -19,6 +24,24 @@ type networkNamespace struct {
...
+	if err != nil && !os.IsExist(err) {
+		panic("Could not create net namespace path directory")
+	}
+}
+
+// GenerateKey generates a sandbox key based on the passed
+// container id.
+func GenerateKey(containerID string) string {
+	maxLen := 12
+	if len(containerID) < 12 {
```
Comment (aboch) [inline sandbox/namespace_linux.go]: Minor one:    < maxLen

File: sandbox/sandbox_linux_test.go [hunk 2]
```diff
@@ -54,7 +54,8 @@ func newInfo(t *testing.T) (*Info, error) {
 	intf.Address = addr
 	intf.Address.IP = ip4
 
-	ip6, addrv6, err := net.ParseCIDR("2001:DB8::ABCD/48")
+	// ip6, addrv6, err := net.ParseCIDR("2001:DB8::ABCD/48")
```
Comment (aboch) [inline sandbox/sandbox_linux_test.go]: Do you may want to keep the commented line ?

File: sandbox/sandbox_linux_test.go [hunk 3]
```diff
@@ -63,7 +64,8 @@ func newInfo(t *testing.T) (*Info, error) {
 
 	sinfo := &Info{Interfaces: []*Interface{intf}}
 	sinfo.Gateway = net.ParseIP("192.168.1.1")
-	sinfo.GatewayIPv6 = net.ParseIP("2001:DB8::1")
+	// sinfo.GatewayIPv6 = net.ParseIP("2001:DB8::1")
```
Comment (aboch) [inline sandbox/sandbox_linux_test.go]: same here

File: sandbox/namespace_linux.go (hunk 1 above)
Comment (mrjana) [inline sandbox/namespace_linux.go]: Ok will change that.

File: sandbox/sandbox_linux_test.go (hunk 2 above)
Comment (mrjana) [inline sandbox/sandbox_linux_test.go]: Yeah that was intentional. wanted to keep that just for reference :-)

File: network.go [hunk 4]
```diff
@@ -387,6 +451,60 @@ func (ep *endpoint) SandboxInfo() *sandbox.Info {
...
+	sinfo := ep.SandboxInfo()
+	if sinfo != nil {
+		for _, i := range sinfo.Interfaces {
+			err = sb.AddInterface(i)
+			if err != nil {
+				return "", err
+			}
+		}
+
+		err = sb.SetGateway(sinfo.Gateway)
```
Comment (aboch) [inline network.go]: SetGateway() does not check for sinfo.Gateway == nil or len(sinfo.Gateway) == 0.
Given sandbox is an independent package, I am not sure we can always assume the v4 gateway is there.

Comment (mrjana) [inline network.go]: Makes sense. Will add a check in `SetGateway()`

File: drivers/bridge/bridge.go [hunk 5]
```diff
@@ -353,15 +339,16 @@ func (d *driver) CreateEndpoint(nid, eid types.UUID, sboxKey string, epOptions i
 	}
 
 	// Create and add the endpoint
+	n.Lock()
 	endpoint := &bridgeEndpoint{id: eid, config: epConfig}
-	n.endpoints[sboxKey] = endpoint
+	n.endpoints[string(eid)] = endpoint
```
Comment (aboch) [inline drivers/bridge/bridge.go]: Would it make the code cleaner to change endpoints key type to be UUID

Comment (mrjana) [inline drivers/bridge/bridge.go]: Ok let's do it.

**Comment by aboch** [comment]
LGTM

**Comment by squaremo** [comment]
> As mentioned earlier in other threads, the idea is to keep the endpoint lifetime independent of containers.

Is there still a useful interpretation for "Endpoints are garbage collected when they no longer belong to any Sandboxes."? (https://github.com/docker/libnetwork/blob/master/ROADMAP.md#concepts)

**Comment by mrjana** [comment]
@squaremo I think not and didn't realize that this wording existed in `ROADMAP.md`. We should fix it as part of this PR. I will take care of it. Thanks for pointing this out.

**Comment by mavenugo** [comment]
@mrjana Thanks for making this change. It is important to have the life-cycles decoupled. LGTM.

**Comment by squaremo** [comment]
I anticipate a couple of problems with this design.

Firstly, it appears to me that Join and Leave create and destroy sandboxes (i.e., the network namespace) respectively, and methods on endpoint are an odd place to put these side-effects. In particular, it's going to be tricky to run a container that has no endpoints, which is a perfectly reasonable thing to do.

Secondly, resources are committed at endpoint creation time rather than when a container is actually run. This commits resources before they are needed (or even without them being needed), and in the case of e.g., IP addresses, makes it more likely that scarce resources will be exhausted inadvertently by the user. It makes more sense that the contract of Join/Leave is to make sure resources are committed and decommissioned, and for the drivers to be invoked to ensure this.

Surely it would be better to have the sandbox creation be explicit (and separate), for Join and Leave to only add and remove interfaces to a sandbox at the time they are needed; and for Join/Leave to be reflected in the driver SPI.

**Comment by mrjana** [comment]
@squaremo Actually the sandbox is created only on the first Join that the container has on any endpoint and is destroyed only when the container leaves the last endpoint. And when you say "to run a container that has no endpoints" I am assuming you are talking about the equivalent of running a container with `--net=none`. This is already supported by the null driver which got committed in #100 so container if it wants to run without any network connectivity it will run. So docker core will make sure that before such a container is run it will create an endpoint on a network in null driver and therefore creates a sandbox with only the loopback interface in it for example. This is all done without complicating libnetwork or docker core for just only a few bytes of overhead.

Secondly this premise of committing certain resources during endpoint creation time is to facilitate the endpoint as a service model. With this model any client container can reach a service with just the endpoint name regardless of which container is providing the service. This facilitates workflow such as someone wants to test out a dev version of their software so they can swap out a prod version with a dev version completely transparently to the clients and then put back the prod version when testing is done. It makes possible sophisticated load balancing architectures possible. It also facilitates IP address and port binding stability when a container moves from one host to another. Host movement for the container is simply done container leaving the endpoint in a given host and re-joining in a different host.

I am aware of the downsides of this such as user not removing the container and therefore locking in some network resources but this is something that we need to educate the users about. That a container not removed might hold up some network resources etc.

Nevertheless, I can assure you that Join/Leave hooks __will__ be provided in the driver api so the driver can optimize on lazily plumbing certain things and delay until a Join on the endpoint happens. For example it may decide not to hook the veth pair onto the bridge until a Join happens etc. We just haven't reached there as we are focusing on setting down the northbound api so that we can finish the docker core integration.

**Comment by monadic** [comment]
@mrjana have you looked at the volumes design?  cc @icecrime @crosbymichael @lukemarsden

**Comment by squaremo** [comment]
> Nevertheless, I can assure you that Join/Leave hooks will be provided in the driver api so the driver can optimize on lazily plumbing certain things and delay until a Join on the endpoint happens.

But won't `CreateEndpoint` and `Join` always be called together -- or are you saying that Join will be called "at the last minute", in other words, when the container is actually started; and similarly, Leave will be called when the container stops? If not that latter, it will make no difference to the ability of the driver to do things lazily, since it will still bear no relation to whether the container needs the resources or not.

Actually the ideal situation is one in which the driver chooses when to commit the resources. For example, we(ave) would like in general to allocate an IP address at the time the container is started, and release it when the container stops. However, it's also plausible that a driver (or indeed weave, when given a particular flag) will want to allocate at the time the endpoint is created, so that e.g., a specific container gets the same IP each time it starts.

**Comment by monadic** [comment]
@mrjana you say "I am aware of the downsides of this such as user not removing the container and therefore locking in some network resources but this is something that we need to educate the users about. That a container not removed might hold up some network resources etc".  

You are essentially proposing that users can be educated to release distributed locks on shared resources.  But, this is exactly why giving a distributed lock to a user has come to be seen as an anti pattern....

**Comment by squaremo** [comment]
> Host movement for the container is simply done container leaving the endpoint in a given host and re-joining in a different host.

This implies some rather onerous requirements on the part of drivers -- doesn't this mean, for example, that a plugin will need to be prepared to accept a Join call for which it has not seen a CreateEndpoint call?

**Comment by mavenugo** [comment]
@monadic I don't see how this design element has any impact on the end-user needing to have any knowledge on distributed locks or even shared resources. The entirety of the Join/Leave design is to make sure there is a clean separation between a service life-cycle and a sandbox life-cycle.
Regarding the resources being locked, it is mostly applicable for ip-address & it infact addresses a long-standing request of IP-Address stability.
The rest of the possible resources applied on the sandbox can be lazily executed during the Join call.

**Comment by squaremo** [comment]
> Regarding the resources being locked, it is mostly applicable for ip-address & it infact addresses a long-standing request of IP-Address stability.
> The rest of the possible resources applied on the sandbox can be lazily executed during the Join call.

It's users inadvertently running out of IP addresses that we are mostly concerned about. I understand the IP stability requirement, but it is not a universal requirement (i.e., a MAY rather than a MUST), and there are ways to support it without _mandating_ it.

**Comment by monadic** [comment]
@mavenugo what @squaremo said.

**Comment by mrjana** [comment]
@squaremo I think there seems to be some things that I haven't explained well in my previous reply.

First of all for endpoints which are implicitly created for a container run the `CreateEndpoint` and `Join` will be called together. So when `CreateEndpoint` and `Join` are called together and `Leave` and endpoint `Delete` called together nobody needs to be worried about unnecessary resource blockage when this happens during container start and stop.

But with this model endpoint becomes a first class resource like a container and in fact it takes some of the responsibility of the container which is networking related. So creating an endpoint is akin to creating a service and the driver allocates all the necessary resources needed for providing that networking service. In fact in the future you are likely see UI which creates an endpoint independently without a container (UI looking more like `docker service create`), exposes a set of ports on the endpoint(very similar things as what happens in container with `-P` option) and along with it UI to join a container to this service. This is the main reason for decoupling endpoints from container Joins. 

The IP address stability is mostly a secondary goal which we nicely achieve with this model without doing much.

All of these things are possible because we have a nice set of composable apis such as `CreateEndpoint/Delete` and `Join/Leave`.

**Comment by mrjana** [comment]
@monadic I haven't looked at volumes design. What about it that relates to this conversation?

**Comment by mrjana** [comment]
> This implies some rather onerous requirements on the part of drivers -- doesn't this mean, for example, > that a plugin will need to be prepared to accept a Join call for which it has not seen a CreateEndpoint > > call?

@squaremo Driver will never get a Join for an endpoint that it has never seen

**Comment by squaremo** [comment]
> Driver will never get a Join for an endpoint that it has never seen

Ok put it this way: if `CreateEndpoint("network1", "abc123", options)` is called on two different hosts, are the drivers on those hosts required to return the "same" endpoint? Otherwise, how is this endpoint migration scenario intended to work.

(And to be a broken record, where are these requirements written down?)

**Comment by squaremo** [comment]
@mrjana Thank you for the expanded explanation. I understand how this model requires an IP address at endpoint creation time.

I think it is fair to say this is not a _technical_ requirement, since the IP address is only needed when the interface is moved inside the namespace, and it is only necessary that this is before the container is _started_. So it seems like a sub-requirement of the "independent lifetimes" model you describe.

Another mode of use (more in line with the requirements given in #9983 and in the roadmap) is that endpoints are an ephemeral means of connecting a container to a network; and in this case, a more useful semantic requirement is that the IP address and other resources are allocated on demand, and released when no longer needed. And it seems like the driver SPI could make this possible by leaving to the driver the choice of when resources, including IP addresses, are committed -- so long as they are in place when they are _technically_ needed.

**Comment by tomdee** [comment]
@mrjana  Should I raise an issue to track `I can assure you that Join/Leave hooks will be provided in the driver api`?

I'm very interested in the exact semantics of when these hooks will be called (an.d in fact when EndpointDeleted is called too for implicitly created endpoints. What guarantees are given about this method being called).

On the issue of IP address assignment at endpoint creation time, for Calico, this is desirable since policy is attached to endpoints and is (ultimately) implemented in terms of IP addresses. IP assignment at creation time would also be useful when dealing with very short lived containers. It means we can set up pools of endpoints (with IPs) so that routing information can be distributed before short-lived containers are started.

Having said that, without a public requirements or design doc it's difficult to comment to whether this is the best approach overall.

**Comment by mavenugo** [comment]
@tomdee am working on the driver api for join/leave as we speak. Expect PR shortly.
Also I will update the design.md shortly.

**Comment by tomdee** [comment]
@mavenugo Perfect, thanks.

**Comment by squaremo** [comment]
> I will update the design.md shortly.

Where is this design.md?

**Comment by dave-tucker** [comment]
@squaremo I've added one in https://github.com/docker/libnetwork/pull/104
While not complete, it should at least form a reasonable basis for documentation-based PRs or for contributors to document any features that they are committing.

**Comment by squaremo** [comment]
@dave-tucker Aha! Thanks

**Pull request merged**